Initial work on gadget-ui updates. Gadgets can be added and managed v…

anuzis opened the pull request:
Initial work on gadget-ui updates. Gadgets can be added and managed via a new single pane on the right. All BE and FE tests pass. Integration tests are in progress, pending final UI tweaks from review.

@seanlip : Is this close to the approach you had in mind? I tried to remove specifying panelName from many front-end areas where it should now be hard-coded, while preserving much of the back-end logic that seems likely to remain useful in future iterations.

In terms of UI, if we have a contextService similar to what we discussed before that would allow the ScoreBar to see if it's in Editor-Edit mode as opposed to Player mode, we could use that to preserve most of the dynamic rendering of gadgets within the preview on the right. Alternatively, we might ultimately want to shift to thumbnail icons in any case since larger gadgets like coding consoles won't fit well in a preview regardless of customization args.

I'm also expecting we want to do some fine tuning on the CSS that contains this preview (for example, changing the green bar to a softer color). If this implementation is in line with what you expect, I'll take it the rest of the way with integration tests and give it to you and Amit for any final CSS tweaks.

seanlip commented:
Thanks @anuzis -- I'll take a look at code, but cc'ing @amitdeutsch to verify the UI/UX flow.

seanlip commented on core/templates/dev/head/editor/EditorServices.js:
```diff
@@ -1098,13 +1098,13 @@ oppia.factory('explorationGadgetsService', [
      */
-    canAddGadgetTo: function(panelName, gadgetData) {
-      var visibilityMap = _getGadgetsVisibilityMap(panelName);
+    canAddGadgetTo: function(gadgetData) {
```
Rename: canAddGadgetToItsPanel

anuzis replied:
Updated.

seanlip commented on core/templates/dev/head/editor/EditorServices.js:
```diff
@@ -1244,9 +1246,15 @@ oppia.factory('explorationGadgetsService', [
+      // REFACTORING(anuzis): panelName formerly passed as 2nd arg below.
+      // Temporarily passing from shared encapsulating variable to see if it
+      // works before refactoring further downstream.
+      // UPDATE: Confirmed this works. @sll: What's your feedback on this
```
This is fine. I agree changeListService.addGadget() only needs to take a single arg.

anuzis replied:
Updated.

seanlip commented on core/templates/dev/head/editor/GadgetEditor.js:
```diff
@@ -100,7 +100,7 @@ oppia.controller('GadgetEditor', [
   /**
-   * @param {Object} gadgetEditData is a dict representing the gadget
+   * @param {Object} gadgetDict is a dict representing the gadget
    *    being edited. It has the following keys: gadget_type, gadget_name,
```
and gadget_panel

Though I would just call it 'panel'...

anuzis replied:
Renamed gadget_panel to panel across the codebase, but kept the associated template file named gadget_panel.html since panel.html seemed too ambiguous. The rest of it is more flexibly named panel which should be easier to adapt to the more universal idea of panels we discussed around the new learner UI.

seanlip commented on extensions/gadgets/ScoreBar/ScoreBar.html:
```diff
@@ -1,8 +1,5 @@
-    <div class="title protractor-test-scorebar-title"><[scoreBarTitle]></div>
-    <progressbar class="fill protractor-test-scorebar-progressbar" max="maxValue" value="getScoreValue()"></progressbar>
-  </div>
+    <progressbar class="protractor-test-scorebar-progressbar" max="maxValue" value="getScoreValue()"><span class="scorebar-title protractor-test-scorebar-title"><[scoreBarTitle]><span></progressbar>
```
nit: indentation

anuzis replied:
Updated

seanlip commented on extensions/gadgets/base.py:
```diff
@@ -42,6 +42,11 @@ class BaseGadget(object):
+    # This must match the name of a stationary gadget panel in the skin.
+    # Example: 'bottom'. In the future, this property may offer other options
+    # like 'draggable'.
+    gadget_panel = ''
```
Do we have validation checking that the only allowable choice for this is 'bottom'? If not I think it would be nice to add it.

anuzis replied:
Added validation in extensions.gadgets.base_test

seanlip replied:
This sounds fine to me -- I agree with the judgment call.

seanlip commented:
@amitdeutsch: FYI, you'll need to switch on `SHOW_GADGETS_EDITOR` in `feconf.py` before you can see the new functionality.

@anuzis: took a quick pass, the implementation approach looks good in general. Will take a more detailed look once the full implementation is ready. Thanks!

amitdeutsch commented:
Hi all,

Looking good! Here are my UI comments:

1) In the 'Insert Gadget' window:

<img width="611" alt="screen shot 2015-09-28 at 7 46 51 am" src="https://cloud.githubusercontent.com/assets/1188049/10138844/2c4c193a-65b5-11e5-95c8-fdcb70e69338.png">

-The gadget should probably have a non-codey name, e.g. Score Bar as opposed to ScoreBar
-There is something off with the padding and margins, I made a quick mockup to show what this should look like with correct margins:

![insergadget](https://cloud.githubusercontent.com/assets/1188049/10139033/f9b7ea2a-65b5-11e5-98b0-1dc57c3b05dc.png)

2) In the 'Customize Gadget' window:

<img width="603" alt="screen shot 2015-09-28 at 7 54 17 am" src="https://cloud.githubusercontent.com/assets/1188049/10139093/3b8c4c98-65b6-11e5-8b5e-3faf5d2d43c6.png">

-I would replace 'Gadget Name' with 'Name', since you're already in the customize gadget window
-Why are Title and Name different things? I think the answer is because 'title' is the label that's displayed on the actual scrollbar, whereas the Name is never seen. Can the 'Title' field just be removed, and the 'Name' field be the label that's displayed with the score bar?
-I found it pretty annoying that I had to fill out all of the fields for customizing the score bar, then found out at the end that I couldn't actually create the score bar because I didn't have any parameters, and therefore had to discard my work. I would recommend having nothing but a message saying something like 'Please create a parameter before creating a scroll bar' if there are no parameters available, instead of showing all the fields when it's not possible to create a score bar.
-I would recommend placing the 'In which states should this gadget be visible?' section at the end, since that section expands depending on the number of cards in the exploration, and we don't want any of the options to get buried at the bottom if this section gets long. I would also recommend changing 'states' to 'cards', since I we use that elsewhere in the UI. Finally, I'm a little concerned that this section is alphabetized, because the names of states are arbitrary when it comes to alphabetical order. I wonder if it would be possible to assign some kind of numbering to states, such that the first state is always number 1, the last state is always the last number, and in-between states are numbered appropriately... just a thought.

3) The gadget panel in the editor

<img width="340" alt="screen shot 2015-09-28 at 10 28 35 am" src="https://cloud.githubusercontent.com/assets/1188049/10143177/b0d381a0-65cb-11e5-95a1-cdcab93b3e3e.png">

It took me a while to figure out what was going on here, since this looks quite strange. Here are some changes that I would recommend:
-The empty preview score bar doesn't add anything, I would simply remove it.
-The green background is very strange, I would get rid of it and add numbering instead, similar to the parameters. I would also add alternating light grey background to differentiate the rows (see mockup below)
-The add gadget button should be at the bottom of the gadget list, similar to the parameters
-The spacing should be even between the panels (i.e. the dark grey boxes in the mockup below should be equal height)
-Finally, clicking the 'pencil' button for the score bar brings up a totally wonky mini window inside of the panel. Clicking the pencil should simply bring up the customization window.

![gadgetspanel](https://cloud.githubusercontent.com/assets/1188049/10143912/aa3194aa-65cf-11e5-8b4c-5fdace2537a1.png)

Note that I added some wishful thinking to the parameters panel, since it's a mess right now.



That's what I've got for now, overall great job on this and let me know if you need CSS help with any of these things.

seanlip commented:
Hi @amitdeutsch, a quick note on name vs title:

> -Why are Title and Name different things? I think the answer is because 'title' is the label that's displayed on the actual scrollbar, whereas the Name is never seen. Can the 'Title' field just be removed, and the 'Name' field be the label that's displayed with the score bar?

Gadget name is basically analogous to state name, and is used to uniquely identify the gadget instance within the exploration. The title, on the other hand, is the optional label that is shown in the score bar.

Should we change 'title' to 'label' instead?

amitdeutsch commented:
My question is why we need both a name and a label. Why not only have one field (called name or title or label), and have that field identify the gadget, and also have that field appear as the label for the gadget?

seanlip commented:
Technically, there may be a case where the labels for different scorebars are the same (say, "Score"). This could be a possibility in the future, if, one day, we e.g. add the ability to specify the scorebar's colour. The green scorebar may represent progress and the yellow one may represent number of insights obtained, or something.

In general, I think that it's possible that a creator may want a score bar's label to be empty, or to be different in some way from what they think of it as (since the score bar's label is more of a surface-level thing). It's basically the difference between the scorebar's unique internal identifier, and something like, say, its colour. The internal identifier can be completely arbitrary, and the creator ought to have a mental model that its value doesn't really matter and it isn't shown to learners (similar to state names). Reusing the identifier as a visible property would break this mental model, though I suppose we could use it as a reasonable default for the label.

amitdeutsch commented:
But then everyone in the real world has to maintain two names for each score bar in order to enable the possibility that someone in the future might have two score bars with the exact same label but different colors. I'm not quite convinced that there is actually a use case where this is needed. I also don't see when an author would want an unlabeled score bar.

The mental model that people develop will arise from the one we create. I think having a single name will make it very easy to create a mental model, whereas having to maintain an internal identifier vs. an external label sounds like a technical detail that isn't actually necessary for a normal user to deal with.

seanlip commented:
OK; I'm fine with omitting the extra field for the moment if @anuzis is -- if needed, it should be fairly straightforward to reinstate it later.

anuzis commented:
@amitdeutsch : Thanks for the great feedback overall. I'm in agreement on pretty much all points.

Regarding gadgets name Vs title: I agree this can be better disambiguated for the ScoreBar, however, I see value in preserving the option for a gadget to have both. The primary purpose for having a name was less for the ScoreBar, and more necessary for gadgets like the AdviceBar which is a fairly generic and likely to inform requirements for similar counterparts in the future.

By default gadgets don't need any customizations. Example: a calculator probably has no title, and might have no customizations at all. Authors simply choose the cards where it should show up.

The AdviceBar is an example where we wanted to support using multiple AdviceBars at different points in an exploration. Example: consider a computer science curriculum that tackles a computational thinking problem. An author might use 3 AdviceBars throughout the Exploration: one offering supplementary tips for getting started (e.g. for learners who may be missing some background that most learners are expected to have and shouldn't be displayed in the primary card), a later one offering extra analogies on breaking a problem down into smaller pieces, and a third that offers a quick reference on using For-loops, Control flow, etc, shown while a learner is trying to implement a solution. We don't want the editor side panel to say "AdviceBar" 3 times, requiring authors to click into each to pull up the customization window to see the tips contained within. By default we name them "AdviceBar", "AdviceBar1", "AdviceBar2", but the expectation is authors will rename gadgets for easier organization during editing: "Getting Started", "Problem Analysis", and "Code Reference".

Like ScoreBars, AdviceBars offer a learner facing "Title" which can be customized. You might want the title to be "Tips" for all cases. In other contexts, a title like "Quick Reference" might feel more suitable. Example screenshot showing a "Tips" title in an AdviceBar containing 1 tip. The gadget is designed to support up to 3 tips which open in HTML modals which can include embedded images, etc.: 
![advicebar](https://cloud.githubusercontent.com/assets/1622682/10156421/d6149fb2-66bb-11e5-89e0-0c5cd125936f.png)

I agree the current UI can be improved, and get the sense the root issue is the UI not separating the gadget's author facing name intuitively enough from the learner-facing customizations.

If we...
1) simplify "Gadget Name" to "Name"
2) add light-gray supporting text "*(Not visible to learners.)*"
3) more visually separate the name from learner-facing customizations

...I think we can strike a balance that intuitively meets the expected use cases.

I like your mockup that shows gadget name instead of attempting to render a preview or thumbnail of each type. If that list of names contains the author-facing names we're likely in good shape... it might be nice if each row retains a name of the original gadget's type as well, e.g. "Getting Started (Advice Bar)", but in the case of the default name it looks weird, e.g. "Advice Bar (Advice Bar)". Perhaps names only is the way to go... or show the rename in parenthesis only if it has been renamed, e.g. "AdviceBar (Getting Started)", "Advice Bar (Problem Analysis)".

What do you think?

amitdeutsch commented:
Michael, I see what you're saying for the AdviceBar, so for me the questions are:

-If AdviceBar did have modifiable labels, are there any other gadgets that would benefit from this? In other words, should every gadget have to specify both a name and a label, even if the only gadget that might benefit from this setup is AdviceBar? I wonder if the label can just be a field specific to AdviceBar; however, I'm actually not sure that even this is needed, as per the point below:

-Is a modifiable label actually necessary for AdviceBar? Is the added complexity worthwhile in order to add the option of labeling your AdviceBar 'Tips' vs. 'Advice'? Or would it be fine for each AdviceBar to have a single, un-changeable label, such as 'Tips'? I suspect that it won't be a problem if each AdviceBar had the same label.

If nothing has a label field, then all you need to worry about is providing a name for each gadget. I know I had recommended using the Name field for as the label for the ScoreBar, but I just had an even better idea– I think the parameter that the ScoreBar is tied to should be the label of the ScoreBar. For example, if the ScoreBar is tied to a parameter called 'Customer Satisfaction', then that should be the label on the ScoreBar. That's how it works in Scratch when you choose to make a variable visible. What do you think of this?

Regarding your last suggestion, I like the idea of displaying the name of the gadget in parentheses (unless it's the default name). I would flip it so that you actually see the name of the gadget first:

1. Advice Bar (Getting Started)
2. Calculator
3. Score Bar (Customer Satisfaction)
4. Score Bar (Money)

How does that sound?

seanlip commented:
> If nothing has a label field, then all you need to worry about is providing a name for each gadget. I know I had recommended using the Name field for as the label for the ScoreBar, but I just had an even better idea– I think the parameter that the ScoreBar is tied to should be the label of the ScoreBar. For example, if the ScoreBar is tied to a parameter called 'Customer Satisfaction', then that should be the label on the ScoreBar. That's how it works in Scratch when you choose to make a variable visible. What do you think of this?

I'll let @anuzis respond to the other points (I'm pretty open either way), but I think that the parameter name for the ScoreBar should not be the label of the ScoreBar. Parameters are akin to variables in a computer program; they are internal identifiers. I don't want someone to use a parameter throughout the exploration, then create a ScoreBar gadget at the end, find that the parameter name is being shown as the label, and then have to correct every instance of this parameter in the entirety of the exploration.

Note that if we had functionality for renaming parameters, these concerns might dissipate -- but currently we don't, and it'll take a while to introduce it. So I lean fairly strongly towards not tying these two things together.

amitdeutsch commented:
I didn't realize there was no way to rename a parameter– thanks for pointing that out. I do think using the parameter name would be the best situation but it sounds like this will have to wait until a future time.

anuzis commented:
Let's treat names and labels as two separate issues. When we agree on a position for each we can find the right UI based our decision.
Issue 1) Should all gadgets have an editable, author-facing name?
Issue 2) Should all gadgets default to having no labels or customizations, but provide the flexibility for gadget developers to offer a customizable "label" option if it enhances the experience?

The first issue is all about the author-facing experience. The second issue is primarily about the learner facing experience and gadget flexibility in general, with a "label" being one of many possible customizations a gadget developer might provide (akin to an Interaction developer deciding how many lines a text input box should be, or what placeholder text should be).

Re - Issue 1: It sounds like we're in agreement the answer to this is yes based on the AdviceBar example. If we prefer to think about Coding Consoles as a more useful gadgets than AdviceBars, a similar benefit of having customizable names applies there as well: imagine two coding consoles that each start with different placeholder text comments. Knowing which is which will be important when authors want to adjust settings or visibility across cards.

Re - Issue 2: I agree it's unnecessary for all gadgets to require labels. A Calculator is a good example with no label. I also agree we could drop the label for both AdviceBars and ScoreBars if that's what we ultimately decide. AdviceBars can always be "Tips", and ScoreBars can always be "Score", or have no label at all. I agree there's not much difference between "Tips" and "Advice", and the difference between "Tips" and "Quick Reference" might not be significant enough to bother allowing authors to customize it.   For ScoreBars, I see a benefit in allowing authors to provide a label, but if necessary authors could describe what the ScoreBar represents on a card.

Overall: I agree it's a judgement call whether either gadget warrants having a customizable learner-facing label. I'm happy going with your recommendation on each. If either gadget (or any future anticipated gadget) offers a label, it brings us back to the original UI question of authors wondering "What's the difference and why have both?"  To that question, if we can visually distinguish the name not being learner facing we may be in good shape.

*I like the idea of displaying the name of the gadget in parentheses (unless it's the default name). I would flip it so that you actually see the name of the gadget first: Advice Bar (Getting Started)  How does that sound?*   Confirming I like this too. I'll implement this.

seanlip commented:
Since we need to bring gadgets in fairly soon, I'm going to say:
  - for issue 1, the answer is yes;
  - for issue 2, let's drop the label on the score bar for now, and just use its name.

@anuzis: I suggest implementing with these assumptions. If anything changes, I'll be responsible for fixing that prior to release. Please let me know if there's anything else you need!

amitdeutsch commented:
@anuzis I agree with your reasoning, and @seanlip I agree with your answers, let's definitely keep an editable, author-facing name for each gadget, and let's use that name as a label for the score bar. We can revisit this if it turns out there are situations where this is too limiting.

anuzis commented:
Sounds good! In 7 meetings today, but blocked 5 hours to implement this tomorrow. Overall: several nice UI improvements -- thanks again for reviewing!

anuzis commented:
**Lots of UI improvements are in!**

**Updates**
- Replaced gadget previews with a gadget's name. Gadgets are listed by their type first, followed by their name in parenthesis if they have been renamed. Example: "Score Bar (Coconuts Surplus)"
- Gadget renaming is now done within the gadget customization window and resembles the UI interaction for renaming states.
- Removed "title" customization argument for ScoreBar & AdviceBar, using the gadget's name for ScoreBars, and a hardcoded "Tips" for AdviceBars.
- Front-end tests updated to pass. Back-end tests required no changes. Integration tests need to be updated, but couldn't due to ChromeDriver being out of date. I'll followup on this.

**UX Polish**
- Adding gadgets shows non-codey names (e.g. 'ScoreBar' is now 'Score Bar'). This applies consistently for gadget name generation as well.
- Refined padding within gadget selection modal.
- Replaced 'Gadget Name' with 'Name' on customize gadget window.
- Moved gadget visibility checkboxes to bottom of the customization form.
- Added a prominent red warning if no valid parameters are available. (The initial suggestion was not displaying other parts of the form at all; wasn't sure how to do this. With the new warning appearing above the visibility settings it's easier to see up front. What do you think?)
- Removed green bar and hidden gadgets dropdown.
- Moved add gadget button to bottom of list, matching parameters side panel.
- Matched spacing between sidebar panels (dark gray boxes mockup have equal height)

**Didn't change**
- Gadget visibility checkbox sorting order from alphabetical to card sequence. Agree this would be an improvement, but not sure how to best implement. The original UI mocks used the visual state-graph with checkboxes for gadget visibility, which would probably be best long term. Turned out that wasn't a viable route yet. Perhaps alphabetical can do for now.
- Striped rows in the gadget sidebar panel. Appears this HTML/CSS is best done via a table which appeared funky in my attempts. The current white panels look clean, but time permitting I agree it's a nice touch to stripe them. Both side panels should probably be striped at the same time.

**Newly Emerged Polish Issues to Flag**
- Slightly awkward section breaks on gadget customization window, appear to be native to the formSchema. Any suggestions how to remove the last line?
![newgadgetcustomizationmodal](https://cloud.githubusercontent.com/assets/1622682/10275515/01cebb94-6b83-11e5-8fba-58fe73ff8ee2.png)

- Gadgets and parameters side panels use different pencil & X icons. Parameters have a small box around them, gadgets don't. The prior review mockups are gadgets style for both (no box). Should we remove the box from parameters editing?
![sidepanels](https://cloud.githubusercontent.com/assets/1622682/10275557/313a15fe-6b83-11e5-8f3d-e135b7dfcb63.png)

Remaining work:
- Getting Gadget Edit & Delete icons to show up horizontal to each gadget. This is probably easy for a CSS expert, but was surprisingly difficult for a non-CSS expert.
- Integration tests need to be updated, but couldn't be run due to ChromeDriver being out of date.

amitdeutsch commented:
Looking good! So, regarding the issues:
-The border at the bottom of the modal is actually in oppia.css, in .oppia-gadget-state-visibility-panel
-Let's stay with the icons you're currently using, since the parameter buttons need to be changed anyhow.
-The icons are overlapping for two reasons; one, they're in their own div, instead of being in the same div as the oppia-gadget-list-item, and two, the .oppia-gadget-list-container class needs to have a 'position: relative' in order to make the 'position: absolute' work correctly for .oppia-gadget-edit-icon and .oppia-gadget-delete-icon.
-Finally, when you click on the pencil icon next to the gadget name in the modal, the edit window looks bizarre.

I fixed all of these issues in a pull request here:
https://github.com/anuzis/oppia/pull/1/files

seanlip commented on core/templates/dev/head/components/gadget_panel.html:
```diff
@@ -1,6 +1,6 @@
-  <div ng-repeat="gadgetSpec in panelContents()">
-    <oppia-gadget class="protractor-test-<[gadgetSpec.gadget_type]>-gadget" gadget-type="gadgetSpec.gadget_type" gadget-customization-args="gadgetSpec.customization_args" show-in-states="gadgetSpec.visible_in_states">
+  <div ng-repeat="gadgetInstance in panelContents()">
+    <oppia-gadget class="protractor-test-<[gadgetInstance.gadget_type]>-gadget" gadget-type="gadgetInstance.gadget_type" gadget-name="gadgetInstance.gadget_name" gadget-customization-args="gadgetInstance.customization_args" show-in-states="gadgetInstance.visible_in_states">
```
Should the protractor-test-* class refer to the name instead of the type? (Since names will be unique, but types won't be?)

anuzis replied:
The problem with using gadget_name is if the name has spaces it ends up breaking into multiple classes. Example: protractor-test-Score Bar-gadget 

Do you know if there's a clever way around this? I touched on this being a limitation of gadget integration testing in the review with Jacob. It's possible the integration tests need to focus on at most one instance of a given gadget type visible on any card.

In prior integration tests it was possible to specify both a gadget_type and panel_name to uniquely identify a gadget instance by its panel, but since a gadget's panel is now hardcoded as part of the gadget we don't have to worry about that use case anymore.

The way I'm thinking about it:
- We catch cases of invalid settings in both front-end and back-end validation (e.g. too many gadgets visible in a panel at the same time)
- Primary role of gadget integration tests may be visual & interactive aspects of specific gadget instances and the ability to add/edit/rename/delete them.

seanlip replied:
Ah, good point.

I think, long term, we can do something like adding a custom protractor-gadget-name attribute to the tag. But that introduces a new pattern, so let's not do it now.

Instead, in the interim, I'd just add a note along these lines:
```
   <!--
     TODO(anuzis): it's possible for multiple gadgets in an exploration to have the
     same type. In such cases, the protractor class used here does not uniquely
     identify the gadget. Find a way to fix this.
   -->
```
and leave it at that for now.

anuzis replied:
Added!

seanlip commented on core/templates/dev/head/editor/EditorServices.js:
```diff
@@ -340,13 +340,12 @@ oppia.factory('changeListService', [
         cmd: CMD_ADD_GADGET,
         gadget_dict: gadgetData,
-        panel_name: panelName
+        panel_name: gadgetData['panel']
```
nit: prefer gadgetData.panel (similar to elsewhere)

anuzis replied:
Confirming refactored all usage of panel_name to gadget_panel consistently across the backend.

seanlip replied:
Oh, actually, I was referring to the `.panel` notation, as opposed to the `['panel']`.

anuzis replied:
Oh, okay. I'll make one more commit using .panel consistently across front-end. I like keeping all the panel_name variables on the backend as the simpler 'panel'. I reviewed that the shorter name always makes sense in context. It additionally sets up more intuitive naming for the future notion of gadgets having a 'draggable' panel, instead of a 'draggable' panel_name.

seanlip commented on core/templates/dev/head/editor/GadgetEditor.js:
```diff
@@ -146,9 +120,36 @@ oppia.controller('GadgetEditor', [
         $scope.gadgetDict = gadgetDict;
         if (gadgetDict.gadget_type) {
           _initializeCustomizationArgs(gadgetDict.gadget_type);
+          gadgetDict.panel = GADGET_SPECS[
```
Here and below: prefer `.panel` to `['panel']`

anuzis replied:
Updated

seanlip commented on extensions/gadgets/ScoreBar/protractor.js:
```diff
@@ -18,6 +18,7 @@
 
 var objects = require('../../objects/protractor.js');
 
+// REFACTORING(anuzis): use name instead of title
```
Can this be fixed?

anuzis replied:
Fixed as part of today's integration tests update.

seanlip commented on extensions/objects/templates/parameter_name_editor.html:
```diff
@@ -1,5 +1,5 @@
 <span ng-model="$parent.$parent.$parent.$parent.localValue" ui-validate="'validate($value)'">
-  <span ng-if="availableParamNames.length === 0">
+  <span style="color:red" ng-if="availableParamNames.length === 0">
```
nit: add space after :

anuzis replied:
Updated

seanlip commented on extensions/skins/skin_classes.py:
```diff
@@ -258,7 +258,6 @@ class ConversationV1(BaseSkin):
     }]
 
-
 class SnapshotsV1(BaseSkin):
```
nit: two lines above top-level classdef

anuzis replied:
Updated

seanlip commented:
With regards to code, this looks very good. All my comments are minor.

@anuzis: I think that if you can fix these and merge @amitdeutsch's changes, we should be all set!

seanlip commented:
Oh, and to address some of the other issues:

> Integration tests need to be updated, but couldn't be run due to ChromeDriver being out of date.

@jacobdavis11: can we fix this?
@anuzis: I think you found a way to get around this temporarily; if they pass under those circumstances then I think it's still fine to merge.

> The initial suggestion was not displaying other parts of the form at all if no valid param exists; wasn't sure how to do this. With the new warning appearing above the visibility settings it's easier to see up front. What do you think?

I think this sounds fine. (Note that I've only looked at the code and not tested the UI; if @amitdeutsch is fine with the latter then so am I.)

> Gadget visibility checkbox sorting order from alphabetical to card sequence. Agree this would be an improvement, but not sure how to best implement. The original UI mocks used the visual state-graph with checkboxes for gadget visibility, which would probably be best long term. Turned out that wasn't a viable route yet. Perhaps alphabetical can do for now.

Agree this is OK to leave for now. Could you file an issue for this so we don't forget it?

> Striped rows in the gadget sidebar panel. Appears this HTML/CSS is best done via a table which appeared funky in my attempts. The current white panels look clean, but time permitting I agree it's a nice touch to stripe them. Both side panels should probably be striped at the same time.

Ditto -- agree this is OK to leave for now; could you please file an issue?

anuzis commented:
@amitdeutsch -- thanks again for all your help. Gadgets are looking much better overall.

@seanlip -- confirming all review items addressed. I'll open the two new issues as requested momentarily.

anuzis commented:
Apologies, one more minor UI update with screenshots:

I realized a ScoreBar's white label text could still be difficult to read against its background when the ScoreBar's value is low (e.g. 5 of 100). I chose a new CSS background-color that ensures the label is always readable while attempting to complement the new learner view background around it. Before and after screenshots below.

![background-color f5f5f5](https://cloud.githubusercontent.com/assets/1622682/10300841/de8f7b6e-6c36-11e5-9f55-ee38d5380d5f.png)

![background-color 9c9ea7](https://cloud.githubusercontent.com/assets/1622682/10300842/e311c944-6c36-11e5-823b-677be2cc5215.png)

It looks alright to me, but my color theory is admittedly weak and I wouldn't be surprised if a more optimal color is available. FWIW I spent some time contrasting around 15 colors and this was the best I arrived at.

anuzis commented:
Confirming this branch is clear to merge to develop.

jacobdavis11 commented:
The integration tests run fine for me both on develop and this branch. Could you let me know more about the problem with ChromeDriver?

Also, please let us know if you're happy with this spec for enabling advanced features (including gadgets) from the settings tab: https://docs.google.com/document/d/1b2MofOjsjqFsWCnrBRafW0FUeDZ3vLpZ8xj8J35_FQ0/edit#

jacobdavis11 commented:
Just got your other email about ChromeDriver so you can ignore the question here!
